skimage/filters/_sparse.py

```diff
@@ -0,0 +1,69 @@
+import numpy as np
+from ._sparse_cy import _correlate_sparse_offsets
+
+
+def broadcast_mgrid(arrays):
+    shape = tuple(map(len, arrays))
+    ndim = len(shape)
+    result = []
+    for i, arr in enumerate(arrays, start=1):
+        reshaped = np.broadcast_to(arr[(...,) + (np.newaxis,) * (ndim - i)],
+                                   shape)
+        result.append(reshaped)
+    return result
+
+
+def correlate_sparse(image, kernel, mode='reflect'):
+    """Compute valid cross-correlation of `padded_array` and `kernel`.
+
+    This function is *fast* when `kernel` is large with many zeros.
+
+    See ``scipy.ndimage.correlate`` for a description of cross-correlation.
+
+    Parameters
+    ----------
+    image : array of float, shape (M, N,[ ...,] P)
+        The input array. If mode is 'valid', this array should already be
+        padded, as a margin of the same shape as kernel will be stripped
+        off.
+    kernel : array of float, shape (Q, R,[ ...,] S)
+        The kernel to be correlated. Must have the same number of
+        dimensions as `padded_array`. For high performance, it should
+        be sparse (few nonzero entries).
+    mode : string, optional
```

Comment:
@scikit-image/core one Q is whether we want to take *scipy* modes as input and translate to the NumPy modes. My instinct says yes since this is meant to be a "mirror" of `ndi.correlate`.

Does anyone have a time machine so we can undo the divergence between the modes? =P

Reply:
It does seem that we have been using the SciPy mode names within `skimage.filters`, so I agree that we should also do that here for consistency.

Unfortunately we previously made the opposite decision for the `skimage.transform` functions `warp`, `reshape` and `resize`! (`warp` actually uses a [utility to convert names from numpy.pad to ndimage equivalents](https://github.com/scikit-image/scikit-image/blob/59fd9682aad41a1f69a35f05d2a7e15ebdfb8b29/skimage/transform/_geometric.py#L9) so it can `ndi.map_coordinates` in some cases).

Reply:
There are more extension modes in pad than in ndimage, but just supporting the subset corresponding to those in ndimage should be sufficient here. The three that change name from numpy->ndimage are:
pad name | ndimage name
--------------|---------------------
nearest | edge
symmetric | reflect
reflect | mirror

The two that are the same for both are `wrap` and `constant`

Reply:
> Note that although the name reflect is present in both, the behavior is not equivalent!

I know. 🤦 Again: time machine. =P